tinycc: fix for darwin

06kellyjac opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

```
ld: malformed 32-bit x.y.z version number: 0.9.27+date=2022-01-11
clang-11: error: linker command failed with exit code 1 (use -v to see invocation)
```

https://hydra.nixos.org/build/178186865/nixlog/2

---

There may be other issues, especially on aarch64-darwin but I don't have a machine

---

Should resolve #181982

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [X] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

jiegec commented:
Failing on aarch64-darwin:

```
../arm64-osx-tcc -ar rcs ../arm64-osx-libtcc1.a arm64-osx-lib-arm64.o arm64-osx-fetch_and_add.o arm64-osx-va_list.o
make[1]: Leaving directory '/private/tmp/nix-build-tcc-0.9.27+date=2022-01-11.drv-0/source/lib'
clang -o c67-tcc.o -c tcc.c -DTCC_TARGET_C67 -w  -DTCC_LIBTCC1="\"c67-libtcc1.a\""       -Wall -O2 -Wdeclaration-after-statement -fno-strict-aliasing -fheinous-gnu-extensions -Wno-pointer-sign -Wno-sign-compare -Wno-unused-result -Wno-string-plus-int -I.
clang -o c67-tcc c67-tcc.o -lm -lpthread -ldl -flat_namespace -undefined warning
ld: warning: ignoring file /nix/store/w864nsavm6gm97aqy5654dc1dppg7fw8-libSystem-11.0.0/lib/crt1.10.6.o, building for macOS-arm64 but attempting to link with file built for macOS-x86_64
buildPhase completed in 4 minutes 8 seconds
running tests
check flags: SHELL=/nix/store/m6pqblbr77ady86apyl7ickafgprjd9f-bash-5.1-p16/bin/bash VERBOSE=y test
make[1]: Entering directory '/private/tmp/nix-build-tcc-0.9.27+date=2022-01-11.drv-0/source/tests'
------------ hello-exe ------------
FIXME: handle reloc type 9 at 1052 [0x600002248022] to 1050
/nix/store/m6pqblbr77ady86apyl7ickafgprjd9f-bash-5.1-p16/bin/bash: line 1: 84412 Bus error: 10           ./hello
+ ../tcc -vv
tcc version 0.9.27 (AArch64 Darwin)
install: /nix/store/23rkk8fib8fb8ljv5dxqc492jam4n26w-tcc-0.9.27+date=2022-01-11/lib/tcc
include:
  /nix/store/w864nsavm6gm97aqy5654dc1dppg7fw8-libSystem-11.0.0/include
  /nix/store/23rkk8fib8fb8ljv5dxqc492jam4n26w-tcc-0.9.27+date=2022-01-11/lib/tcc/include
libraries:
  /nix/store/w864nsavm6gm97aqy5654dc1dppg7fw8-libSystem-11.0.0/lib
  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/lib
libtcc1:
  /nix/store/23rkk8fib8fb8ljv5dxqc492jam4n26w-tcc-0.9.27+date=2022-01-11/lib/tcc/libtcc1.a
crt:
  -
elfinterp:
  /usr/lib/dyld
+ otool -L ../tcc
../tcc:
        @rpath/libtcc.dylib (compatibility version 0.9.27, current version 0.9.27)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.60.1)
+ exit 1
make[2]: *** [Makefile:87: hello-exe] Error 1
make[1]: *** [Makefile:83: all] Error 2
make[1]: Leaving directory '/private/tmp/nix-build-tcc-0.9.27+date=2022-01-11.drv-0/source/tests'
make: *** [Makefile:427: test] Error 2
error: builder for '/nix/store/sbhsnbrsqxz9dg3ldp3vilai4k5x1ysj-tcc-0.9.27+date=2022-01-11.drv' failed with exit code 2;

```

06kellyjac commented:
![mac](https://user-images.githubusercontent.com/9866621/179949014-3a36749a-bee8-474c-b664-be6e9155fe8f.png)

Success on x86_64 darwin, building deno now to check
Need to double check with aarch64

aiotter commented on pkgs/top-level/all-packages.nix:
```diff
@@ -14344,7 +14344,7 @@ with pkgs;
   };
 
-  tinycc = callPackage ../development/compilers/tinycc { };
+  tinycc = darwin.apple_sdk_11_0.callPackage ../development/compilers/tinycc { };
```
```suggestion
  tinycc = callPackage ../development/compilers/tinycc {
    stdenv = if stdenv.isDarwin then darwin.apple_sdk_11_0.stdenv else stdenv;
  };
```

I'm afraid that it's a good practice to build everything on `darwin.apple_sdk_11_0.stdenv` even when you're not on darwin.

06kellyjac replied:
This is what go1.18 is doing, also note the platform doc in this ripgrep:

```
λ rg "darwin.apple_sdk_11_0.callPackage"
doc/stdenv/platform-notes.chapter.md
66:  `apple_sdk` in your derivation and use `pkgs.darwin.apple_sdk_11_0.callPackage` instead of `pkgs.callPackage`. On Linux, this will
67:  have the same effect as `pkgs.callPackage`, so you can use `pkgs.darwin.apple_sdk_11_0.callPackage` regardless of platform.

pkgs/top-level/all-packages.nix
13457:  go_1_18 = darwin.apple_sdk_11_0.callPackage ../development/compilers/go/1.18.nix {
21829:  buildGo118Package = darwin.apple_sdk_11_0.callPackage ../development/go-packages/generic {
21840:  buildGo118Module = darwin.apple_sdk_11_0.callPackage ../development/go-modules/generic {
```

aiotter replied:
`callPackage` and `stdenv` under `pkgs.darwin.apple_sdk_11_0.callPackage` is quite new and there is no standard way of packaging with it. Maybe I should create an issue and discuss about the best practice? Thanks for letting me know about the document!

06kellyjac replied:
yep. ideally we can just get this merged to unblock deno & have a discussion elsewhere.

The new nixpkgs-architecture team might be interested https://github.com/nixpkgs-architecture

aiotter reviewed (changes requested):
Thank you for working with this! deno is an essential tool for me. I really appreciate your work.

06kellyjac commented:
The deno build took ages but finally finished

![macbuild](https://user-images.githubusercontent.com/9866621/179964985-8724bc9f-c988-4d14-8fa3-19a2d686569a.png)


Running their ffi test bench worked fine (I had to copy our deno into target/debug since that's how it resolves the test dylib)

![macbench](https://user-images.githubusercontent.com/9866621/179964823-5a7e5560-8fc5-4542-9998-a00d7cac7d98.png)

06kellyjac commented:
revert last commit so we have a cached version while the upstream src is broken

can move to GH mirror if we have to in the future

siraben commented:
Result of `nixpkgs-review pr 181990` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages failed to build:</summary>
  <ul>
    <li>deno</li>
    <li>nrpl</li>
    <li>tinycc</li>
  </ul>
</details>

siraben commented:
Seems to fail on aarch64-darwin:
```ShellSession
$ nix log /nix/store/c37pgc1jnpfkpcpalpslay356gajbp5m-tcc-0.9.27+date=2022-01-11.drv | ix
http://ix.io/44YZ
```

siraben commented:
Result of `nixpkgs-review pr 181990` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>nrpl</li>
    <li>tinycc</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>deno</li>
  </ul>
</details>

06kellyjac commented:
Ok. It looks like ofborg was able to pull the src from the cache and failed at the tests so at some point between this commit and the other darwin (at least x86_64) is fixed.

I will probably point it to github's latest commit

jiegec commented:
Maybe we can mark tinycc broken on aarch64-darwin, and remove the tinycc dependency from deno on aarch64-darwin.

06kellyjac commented:
it was working for both (IIRC) on the commit we couldn't consistently pull

In it's current state it's broken for both so we'd need to drop tinycc as a dependency for both.

I tried it but the build was taking ages. The build generally takes ages so I'll try it again but it might not actually work

06kellyjac committed:
tinycc: unstable-2022-07-15 -> 0.9.27+date=2022-05-26

repoOrCz is having issues so we can't fetch that commit
Moved to the latest commit we can reliably fetch from the GH mirror

06kellyjac commented:
Ok yeah dropping tinycc from deno will require quite a hefty patch so fixing this here would be ideal.

06kellyjac committed:
tinycc: fix for darwin

With help from the nix on macOS matrix channel

Co-authored-by: Randy Eckenrode <randy@largeandhighquality.com>

06kellyjac commented:
I mostly expected the aarch64 issue but I managed to build x86_64-darwin so ofborg's failure is a bit odd

no idea where `error: unknown target CPU 'armv8.3-a+crypto+sha2+aes+crc+fp16+lse+simd+ras+rdm+rcpc'` is coming from

there is an `xcrun` error logged but that also logged in the build I did that succeeded

jiegec commented:
> I mostly expected the aarch64 issue but I managed to build x86_64-darwin so ofborg's failure is a bit odd
> 
> no idea where `error: unknown target CPU 'armv8.3-a+crypto+sha2+aes+crc+fp16+lse+simd+ras+rdm+rcpc'` is coming from
> 
> there is an `xcrun` error logged but that also logged in the build I did that succeeded

This error usually comes from building x86_64 binaries on aarch64-darwin, because the toolchain cannot build universal binaries.

AndersonTorres commented on pkgs/development/compilers/tinycc/default.nix:
```diff
@@ -1,50 +1,31 @@
     which
   ];
 
-  pkgconfigItems = [
```
Is this failing?

06kellyjac replied:
I think this was dropped in a bad rebase since im not familiar with this chunk of code...
Ill need to add it back in and check

SuperSandro2000 commented on pkgs/development/compilers/tinycc/default.nix:
```diff
@@ -1,50 +1,31 @@
 stdenv.mkDerivation rec {
   pname = "tcc";
-  version = "unstable-2022-07-15";
+  version = "0.9.27+date=2022-05-26";
```
```suggestion
  version = "0.9.27";
```
The normal version format has no dates. We only do that for the unstable one.

domenkozar commented:
Any updates?
